Research + Fishing  Modding APIs

JPANv2 opened the pull request:
### What is the new feature?
This Pull request creates two new hooks to the Research System:

One "CanResearch",  in order to allow modders to allow or not an item to be researched, regardless of it was already fully researched in the past or even if it has a research value;

One "OnResearched", called before the item is modified, that allows for custom behaviour to be run if an item was either partially or fully researched.

### Why should this be part of tModLoader?
These hooks allow for many features to be implemented by modders that are otherwise pratically impossible to achieve, including:

- Allow Modders to prevent an item from being researched before a certain stage in game is reached (ex- Hardmode Only)
- Allow Modders to force an item to be researched even if its researched value is already full, like researching a weapon a second time to unlock its prefix, or create Random Research items that add research for a random item when researched.
- Allows for code to run when the item is researched, to add features such as my mod's "Research unlocks everything crafted from that item".

### Are there alternative designs?
It would be possible to make the design more complex, by adding new parameters to the functions, but they would add nothing new.

### Sample usage for the new feature
`class ChestOfRandomResearch : ModItem
{

public override bool? CanResearch(){
 return true;
}

public override void OnResearched(bool fullyResearched){
int stack = item.stack;
for(int i = 0; i < stack; i++){
Item newItm = new Item();
newItm.SetDefaults(Main.rand.Next(1, ItemLoader.Count));
newItm.stack = 1;
int amount = 0;
Main.CreativeMenu.SetItembyIndex(newItm,0);
Main.CreativeMenu.SacrificeItem(amount);
}
}
}
`
### ExampleMod updates
The above example could be added to the ExampleMod, after tested for bugs.

JPANv2 committed:
Created CanResearch and OnResearched Hooks for ModItems and GlobalItems

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ItemLoader.cs:
```diff
@@ -425,6 +425,38 @@ private class HookList
+				
+				if (canResearch != null)
+					return canResearch;
+			}
```
all hooks like this need to decide on a precedence behaviour. Returning the result of the first hook to give a non-null (vanilla) response will be weird when one mod says yes and another says no.

You need to decide, and describe in the documentation of the hook, which should take priority.

Do this by suggesting use cases for these hooks.

Solxanich commented:
Hello @JPANv2 ; thank you for drafting this PR.
As the tModLoader project continues to progress, I'd like to ask if you intend to return to complete this PR and address Chicken-Bones's review comments?

If not, could you please let us know, so we can have someone else pick up the branch and bring it to completion?

Thank you for your time.

JPANv2 committed:
Changed Can research to have correct priorities, added those priorities in comments

JPANv2 committed:
Merge branch '1.4' into 1.4

JPANv2 commented:
Created the requested changes and updated my branch to the latest TMod 1.4. However, it now says that my branch can't build, even though checking it out and clean/build in VS works just fine.

JPANv2 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4

JPANv2 committed:
Added CatchFish that uses FishingAttempt to show all fishing data + refs to change itemid, spawn npc, and custom  Sonar Text.

Added ModifyCaughtFish to allow changing the fished item before added to the player inventory.

Added WillConsumeBait to determine if bait should be not consumed, forced consumed or just do vanilla treatment.

JPANv2 commented:
Updated version of both Fishing changes and Research Hooks. Unfortunately, could not separate them into two forks, but both are building, using the latest patch and tested working. If needed, I'll also add an ExampleMod example for both APIs.

JPANv2 committed:
Fixed small bug with CanResearch
Added examples to Example Mod.

JPANv2 committed:
Switched the files back for example Drop Conditions

JPANv2 committed:
fix formatting in exampleDropCondition

JPANv2 committed:
Fixed formatting and added comments to ExampleFishingPlayer

JPANv2 committed:
Fix formatting and renamed hook CanConsumeBait

JPANv2 committed:
Fixed ExampleDrop Swapped descriptions

JPANv2 committed:
fixed more formatting

JPANv2 committed:
Removed a needless repetition in ExampleFishingPlayer

Chicken-Bones commented on ExampleMod/Content/Items/ExampleResearchPresent.cs:
```diff
@@ -0,0 +1,61 @@
+			 accessories*/
+			if (fullyResearched) {
+				for (int i = 1; i < ItemLoader.ItemCount; i++) {
+					Item tr = new Item(i);
```
Use `ContentSamples.ItemsByType` to query item properties, rather than making a new item. And also below on line 47

JPANv2 replied:
didn't know about that function. Will apply it where it's relevant.

Chicken-Bones commented on ExampleMod/Content/Items/ExampleResearchPresent.cs:
```diff
@@ -0,0 +1,61 @@
+				Main.NewText("You got all accessories!");
+				return;
+			} else {
+				int j = this.Item.stack;
```
It's not clear to me exactly what the intended behaviour is here. Needs a comment.
From what I can see, if you research multiple at once, they all get consumed, but you might run out of rolls on the item table (fixed 1000) before that.
If you run out of rolls, you may have actually researched some accessories, but the text will still say "no new accessory was inside"
The `ResearchItem` call doesn't check if it's already been fully researched either...

JPANv2 replied:
I changed ResearchItem to have a out boolean alreadyResearched, that will indicate if something was already researched before being sent out. only place that will not have it is the original Sacrifice function.

Chicken-Bones commented on ExampleMod/Content/Items/ExampleResearchPresent.cs:
```diff
@@ -0,0 +1,61 @@
+		public override void SetStaticDefaults() {
+			DisplayName.SetDefault("Example Present");
+			Tooltip.SetDefault("Contains a random accessory! Try researching it for infinite of it!");
+			CreativeItemSacrificesCatalog.Instance.SacrificeCountNeededByItemId[Type] = ItemLoader.ItemCount;
```
Does sacrifice count changing when you add/remove mods cause a problem? What if you sacrifice 5k of a thing and then unload a mod bringing the count needed down to 4k. Does the item become fully researched then and then become unresearched when you add more mods?

JPANv2 replied:
I changed the Research value from a random integer (1-100) to ItemCount and back, and if the player already researched fully the present, would still be able to research it further because of CanResearch returning always true. Also, when changing the values, the item went from fully researched to only partially researched whenever it went from lower to higher numbers, but researching one new item will reduce the amount researched to the maximum at the time.

Chicken-Bones reviewed (changes requested):
Partial Example review

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/GlobalItem.cs:
```diff
@@ -191,6 +191,34 @@ public abstract class GlobalItem : GlobalType<Item>
 		}
 
+		/// <summary>
+		/// Allows you to choose if a given bait will be consumed by a given player
```
Need to specify precedence here.

JPANv2 replied:
done False trumps true, trumps null. Placed in the comments of both here and player hooks.

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -4180,6 +4180,15 @@
+ 			if (baitTypeUsed == 2673)
+ 				flag = true;
+ 
++			bool? forced = PlayerLoader.CanConsumeBait(this, item) ?? ItemLoader.CanConsumeBait(this, item);
```
You might need to use `CombinedHooks` here, because you need to combine the precedence of the two results.

JPANv2 replied:
This one I had decided that player-based decisions take precedence over item decisions, for scope reasons. But I can change it to always have true as precedence for every single case.

JPANv2 committed:
Made requested changes:
- Fixed Hook priority for CanConsumeBait
- Changed ResearchItem(...) to include an out boolean that tells if the item was already researched.
-Changed examples for this new features.

Chicken-Bones committed:
Cleanup conditions and examples.

JPANv2 committed:
changed canResearch to true by default, added PreResearchItem hook and updated example present to reflect those changes

JPANv2 committed:
Merge branch '1.4' of https://github.com/JPANv2/tModLoader into 1.4

JPANv2 committed:
Added  conveinience property "SacrificeTotal" to ModItem, to simplify adding research values to the system.

JPANv2 committed:
Fixed sonar display

Chicken-Bones committed:
Cleanup CreativeUI.cs.patch

Chicken-Bones committed:
Remove PreItemResearch, use a MonoMod hook instead. Cleaner example comments and helper methods.

Chicken-Bones committed:
Minor patch minimisation

Chicken-Bones committed:
Merge branch '1.4' into fishingpr

JPANv2 committed:
Small fixes to ExamplePresent, to limit its SacrificeTotal to the max allowed, and correctly decrease the stack of the ref item
